solarforecastarbiter/reports/figures.py

```diff
@@ -0,0 +1,211 @@
+    """Make a nice human readable name."""
+    caps = ('ac_', 'dc_', 'poa_', 'ghi', 'dni', 'dhi')
+    fname = variable
+    for cap in caps:
+        fname = fname.replace(cap, cap.upper())
+    fname = fname.replace('_', ' ')
+    return fname + f' ({units})'
+
+
+def line_or_step(interval_label):
+    if 'instant' in interval_label:
+        plot_method = 'line'
+        kwargs = dict()
+    elif interval_label == 'beginning':
+        plot_method = 'step'
+        kwargs = dict(mode='before')
+    elif interval_label == 'ending':
+        plot_method = 'step'
+        kwargs = dict(mode='after')
+    return plot_method, kwargs
+
+
+def construct_fx_obs_cds(fx_obs, fx_values, obs_values):
+    """
+    Construct a tuple of metadata, data for the plot functions.
+
+    Parameters
+    ----------
+    fx_obs : solarforecastarbiter.datamodel.ForecastObservation
+    fx_values : pandas.Series
+    obs_values : pandas.Series
+
+    Returns
+    -------
+    fx_obs : solarforecastarbiter.datamodel.ForecastObservation
+    cds : bokeh.models.ColumnDataSource
+    """
+    data = pd.DataFrame({'observation': obs_values, 'forecast': fx_values})
+    data['timestamp'] = data.index
+    cds = ColumnDataSource(data)
+    return fx_obs, cds
+
+
+def _obs_name(fx_obs):
+    # TODO: add code to ensure obs names are unique
+    name = fx_obs.observation.name
+    if fx_obs.forecast.name == fx_obs.observation.name:
+        name += ' Observation'
+    return name
+
+
+def _fx_name(fx_obs):
+    # TODO: add code to ensure fx names are unique
+    name = fx_obs.forecast.name
+    if fx_obs.forecast.name == fx_obs.observation.name:
+        name += ' Forecast'
+    return name
+
+
+def timeseries(fx_obs_cds, start, end):
+    """
+    Timeseries plot of one or more forecasts and observations.
+
+    Parameters
+    ----------
+    obs_fx_cds : tuple of (ForecastObservation, cds) tuples
+        ForecastObservation is a datamodel.ForecastObservation object.
+        cds is a Bokeh ColumnDataSource with columns
+        timestamp, observation, forecast.
+    start : pandas.Timestamp
+        Report start time
+    end : pandas.Timestamp
+        Report end time
+
+    Returns
+    -------
+    fig : bokeh.plotting.figure
+    """
+
+    palette = iter(PALETTE)
+
+    fig = figure(
+        sizing_mode='scale_width', plot_width=900, plot_height=300,
+        x_range=(start, end), x_axis_type='datetime',
+        tools='pan,wheel_zoom,reset',
+        name='timeseries')
+
+    plotted_objects = []
+    for fx_obs, cds in fx_obs_cds:
+        if fx_obs.observation in plotted_objects:
+            pass
+        else:
+            plotted_objects.append(fx_obs.observation)
+            plot_method, kwargs = line_or_step(
+                fx_obs.observation.interval_label)
+            name = _obs_name(fx_obs)
+            getattr(fig, plot_method)(
+                x='timestamp', y='observation', source=cds,
+                color='black', legend=name,  **kwargs)
```

Comment:
Do you really want this to allow for more than one observation series?

Reply:
If we're plotting multiple forecasts against a single observation, I think they should be in separate CDS. No reason to serialize and send the observation data multiple times (at least I think all the data in the CDS is sent regardless of whether it is used in a plot or not)

Reply:
Wouldn't that be trivial to compress?

Reply:
> Do you really want this to allow for more than one observation series?

https://solarforecastarbiter.org/usecases/#uc1D

Reply:
maybe easy to compress, but it will have to be decompressed client side. No reason to waste memory; this isn't an electron app.

Even for that use case, in a report, wouldn't it make more sense to plot some normalized errors for the sites/aggregates? Otherwise scaling and clutter might obscure the plot

Reply:
> (at least I think all the data in the CDS is sent regardless of whether it is used in a plot or not)

I inspected the script inserted into the demo page. It does have observation, forecast, index, and timestamp for both of the `fx_obs_cds` elements. The repeated arrays have exactly the same string representation.

Reply:
will the observations and forecasts even necessarily have the same index? Probably not for most right? In that case, maybe each observation and each forecast should be have a separate CDS. And the CDS for scatter should have the resampled/matched observations and forecast w/ no timestamp

Reply:
> will the observations and forecasts even necessarily have the same index? 

Paired obs/forecasts within a CDS, yes. Across different CDS, not necessarily.

If we're using different CDS in each plot then maybe we should make them within the plot functions. I'm sure it would be more efficient but it would prevent cross-plot interactivity. I think doing things like highlighting a section of the time series and seeing the scatter plot update dynamically would be helpful if not too expensive.

Reply:
:dizzy_face: what about multiple fx-obs pairs but with different interval lengths but with the same Observation? Only one of the resampled observations series is plotted on the timeseries, no? Maybe this should just take the observation values separately, and the CDS of obs-fx pairs that are the same as scatter (to maintain the linking) for the forecast lines? 

I agree that it would be nice to lasso some points on the scatter and see which ones the correspond to on the timeseries

Reply:
1 year of 5 minute data with 2 observations (~200k points) requires about 5 seconds of processing time on my laptop to create the file. The file is 6.4 MB. It takes a couple of seconds for bokeh to load the page. Firefox memory profiler says the page is 159 MB, though my system profiler suggests ~500 MB. In any case, I think we're in an ok spot for now.

Reply:
> what about multiple fx-obs pairs but with different interval lengths but with the same Observation? 

I assumed that we'd have different Observation objects for that (perhaps they'd be dynamically generated from the true Observation).

> Only one of the resampled observations series is plotted on the timeseries, no?

Disagree. I think it's useful to plot say 1 minute obs and resampled 60 minute average obs on the same timeseries plot.

Reply:
I was referencing
```python
if fx_obs.observation in plotted_objects:
    pass
```
but not an issue if you make different Observations. I think there should be one true Observation and dynamically generate the resampled ones. 

> Disagree. I think it's useful to plot say 1 minute obs and resampled 60 minute average obs on the same timeseries plot.

Should have a palette of greys for the observations then?